### PR TITLE
[Modules] Update DiagnosticSettings

### PR DESCRIPTION
**Pull request by AlexanderSehr**
# Description

- Removed `RetentionPolicy` from DiagnoticSettings declaration as it was deprecated
- Added a test case to the storage account `LifecycleManagement` feature to cover its deployment
- Added dependency of `LifecycleManagement` on `BlobServices` deployment to ensure required settings (if used) are set first.

| Pipeline |
| - |
| [AnalysisServices - Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml) |
| [KeyVault - Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml) |
| [Storage - StorageAccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml) |

# Type of Change

Please delete options that are not relevant.

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: constructs/Compute/virtualMachinesMultiple/main.bicep [hunk 1]
```diff
@@ -230,11 +230,6 @@ param extensionCustomScriptProtectedSetting object = {}
 @description('Optional. Location for all resources.')
```
Comment (AlexanderSehr) [inline constructs/Compute/virtualMachinesMultiple/main.bicep:230]: There are currently a truckload of conflicts due to the file genreations. I'll re-generate all once the review is done.
cc: @ahmadabdalla

File: utilities/tools/Set-ModuleReadMe.ps1 [hunk 2]
```diff
@@ -498,7 +498,6 @@ Order the given JSON object alphabetically. Would result into:
 
 @{
     name: 'carml'
-    diagnosticLogsRetentionInDays: 7
```
Comment (ahmadabdalla) [inline utilities/tools/Set-ModuleReadMe.ps1:501]: Need to update the `.EXAMPLE` to remove the `diagnosticLogsRetentionInDays` examples.

File: modules/healthcare-apis/workspace/README.md [hunk 3]
```diff
@@ -114,7 +114,6 @@ fhirServices: [
         corsMaxAge: 600
         corsAllowCredentials: false
         location: location
-        diagnosticLogsRetentionInDays: 7
```
Comment (ahmadabdalla) [inline modules/healthcare-apis/workspace/README.md:117]: This file still contains the `diagnosticLogsRetentionInDays` in the Parameter usage section for JSON. 

image

File: constructs/Compute/virtualMachinesMultiple/main.bicep (hunk 1 above)
Comment (ahmadabdalla) [reply constructs/Compute/virtualMachinesMultiple/main.bicep:230]: A couple of readmes need to be regenerated:

image

**Review by ahmadabdalla — COMMENTED** [review]
@AlexanderSehr Some minor comments. Other than that all good!! Btw I have tested a deployment of an NSG locally when having the latest API + diag updates and it deploys successfully.

File: modules/healthcare-apis/workspace/README.md (hunk 3 above)
Comment (AlexanderSehr) [reply modules/healthcare-apis/workspace/README.md:117]: removed

File: utilities/tools/Set-ModuleReadMe.ps1 (hunk 2 above)
Comment (AlexanderSehr) [reply utilities/tools/Set-ModuleReadMe.ps1:501]: removed

File: constructs/Compute/virtualMachinesMultiple/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply constructs/Compute/virtualMachinesMultiple/main.bicep:230]: removed

**Comment by j-d-harvey** [comment]
Don't understand how you've closed this. Just ran into it myself and the docs are not correct: https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/diagnosticsettings?pivots=deployment-language-bicep

**Comment by AlexanderSehr** [comment]
> Don't understand how you've closed this. Just ran into it myself and the docs are not correct: https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/diagnosticsettings?pivots=deployment-language-bicep

Hey @j-d-harvey, it was closed due to the review, test and subsequent merge. Could you help me understand what about our docs is incorrect? The general link to the temple reference's template is a bit generic 😉

**Comment by wsucoug69** [comment]
> > Don't understand how you've closed this. Just ran into it myself and the docs are not correct: https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/diagnosticsettings?pivots=deployment-language-bicep
> 
> Hey @j-d-harvey, it was closed due to the review, test and subsequent merge. Could you help me understand what about our docs is incorrect? The general link to the temple reference's template is a bit generic 😉 Note that we are not owning the Azure Template References website and merely link to it. It is owned by a different party inside Microsoft and is generated based on the API specs which are still correct. You "can" still deploy the retention policy with a value of 0 days if I'm not mistaken. I wouldn't recommend it though. For the website to correctly update, it would require a new API version which I would expect to be released soon...

I am so confused as how this is not a new bug? I am referencing the latest API version diagnosticSettings@2021-05-01-preview.

The documentation here: https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/diagnosticsettings?pivots=deployment-language-bicep is current. When looking at the (latest) api version it says:  

retentionPolicy: days | the number of days for the retention in days. A value of 0 will retain the events indefinitely. | int (required)

We are setting a value of 30 which has been working in our service for over 7 months until our most recent deployment (today).

This to me is still a bug and should be open correct?

**Comment by AlexanderSehr** [comment]
Hey @wsucoug69,
 it is indeed a new issue - on the Diagnostic Settings side we're deploying. The user that opened the linked issue, added this [link](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/migrate-to-azure-storage-lifecycle-policy) as a reference. From what it looks like, the RetentionPolicy is deprecated across the board (for all API versions). While this appears to have been announced for some time - one doesn't notice it during deployments just until recently when deployments started failing.

Because this is a very bad customer experience (why else have API versioning), I also opened a new [feature request](https://github.com/Azure/bicep/issues/11728) on the Bicep-side to enable a 'warning' also via the template.

Long story short - every user of the DiagnosticSettings that deployed the `retentionPolicy` property up until now has to change their code (similar to how we did it in this PR). Going forward, the retention times are configured on the StorageAccount-side. So it's not a 'bug' but rather by-design - just many people (including me) didn't know about it until our templates started failing.
